File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    
    left = [1,4,7]
    right = [3,6,9]
    hand_L = -1
    hand_R = -2
    hands = ''
    key = [[1,2,3],[4,5,6],[7,8,9],[-1,0,-2]]
    
    for i in range(len(numbers)):
        if numbers[i] in left : 
            hands += 'L'
            hand_L = numbers[i]
        elif numbers[i] in right : 
            hands += 'R' 
            hand_R = numbers[i]
        else:
            for r in range(4):
                for c in range(3):
                    if key[r][c] == numbers[i] : tx,ty = r,c
                    if key[r][c] == hand_R : rx,ry = r,c
                    elif key[r][c] == hand_L : lx,ly = r,c
            R_d = abs(tx-rx) + abs(ty-ry)
            L_d = abs(tx-lx) + abs(ty-ly)
            if R_d > L_d:
                hands += 'L'
                hand_L = numbers[i]
            elif R_d < L_d:
                hands += 'R'
                hand_R = numbers[i]
            else:
                if hand == 'right':
                    hands += 'R'
                    hand_R = numbers[i]
                else:
                    hands += 'L'
                    hand_L = numbers[i]
    return hands

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    let answer = '';
    const keypad = {
        1 : [1,1], 2 : [1,2], 3 : [1,3],
        4 : [2,1], 5 : [2,2], 6 : [2,3],
        7 : [3,1], 8 : [3,2], 9 : [3,3],
        "*" : [4,1], 0 : [4,2], "#" : [4,3],
    }
    let lSpot = keypad["*"], rSpot = keypad["#"] 
    numbers.forEach((num,idx) => {
        if ([1,4,7].includes(num)) {
            lSpot = keypad[num]
            answer += "L"
        }
        else if ([3,6,9].includes(num)) {
            rSpot = keypad[num]
            answer += "R"
        }
        else if ([2,5,8,0].includes(num)) {
            const lDistance = Math.abs(keypad[num][0] - lSpot[0]) +  Math.abs(keypad[num][1] - lSpot[1])
            const RDistance = Math.abs(keypad[num][0] - rSpot[0]) +  Math.abs(keypad[num][1] - rSpot[1])
            const Distance = lDistance - RDistance
            if (Distance < 0){ 
                lSpot = keypad[num]
                answer += "L"     
            }
            else if (Distance > 0){
                rSpot = keypad[num]
                answer += "R"
            }
            else {
                if (hand === "left"){
                    lSpot = keypad[num]
                    answer += "L" 
                }
                else if (hand === "right"){
                    rSpot = keypad[num]
                    answer += "R"
                }
            }
        }
    })
    
    return answer;
}